### ISSUE TITLE
[wgsl] Offset decorations on Structs

### ISSUE THREAD

**Issue by dj2** [gpuweb/gpuweb#561]
With WGSL currently the `[[offset x]]` decoration is required for `SSBO` and `UBO` structs. This makes it simpler for a WGSL compiler as we're told the offsets to use, but makes it harder for a user as you have to figure out the offsets to use.

There are, at least, four ways to specify the offsets in Vulkan. `std140`, `std430`, `relaxed block layout` and `scalar block layout`. In all cases they turn into SPIR-V with specific offsets provided.

The layout requirements depends on the of buffer in vulkan (modulo extensions which change it) so that uniforms at std140 and ssbos are std140 or std430 I believe.

Do we leave WGSL as is and require explicit offsets, or do we relax and allow a given layout to be specified?

**Comment by dj2** [comment]
Being explicit removes the ability for a lot of potential bugs in implementations that convert to SPIR-V. The rules to convert to `std140` and `std430` are tricky and can be difficult to implement.

**Comment by magcius** [comment]
The only vendor without support for scalar layouts is (IIRC) Broadcom. So you'd lose the ability to run WebGPU on rpi if you required support scalar layout. Up to committee whether that's important enough of a target.

[[offset]] doesn't solve alignment of arrays. The classic packing rules of DX10/DX11 say that array elements are aligned to 16-byte offsets. So `float xyz[2];` is 32 bytes. I can't remember if DX12 or scalar layout change things here.

**Comment by Kangz** [comment]
Support for [VK_EXT_scalar_block_layout](https://vulkan.gpuinfo.org/displayextension.php?name=VK_EXT_scalar_block_layout) doesn't look great. I've been wondering for a while if a smart enough compiler could translate scalar layout code into std140 / std430 by losing all typing information and looking at the buffer as a flat array of u32, but we shouldn't rely on it for WebGPU.

The packing rules for DX10 / DX11 that you linked seem to be exactly the same as the `std140` rules.

I don't have an otherwise strong opinion on what to do for WGSL.

**Comment by litherum** [comment]
On the other hand, no existing shading languages require authors to provide these offsets. It seems difficult to claim the language is human writable if we require authors to do something substantial they don't have to do literally anywhere else. Especially if there are only one or two acceptable layouts, and if authors type anything else other than the correct number for any field in their struct, they'll get a compile error.

**Comment by litherum** [comment]
@Kangz 

> looking at the buffer as a flat array of u32, but we shouldn't rely on it for WebGPU.

I thought you were already relying on this for HLSL.

**Comment by Kangz** [comment]
Oh that's right, we kind of do that with `RWByteAddressBuffer` for storage buffers, because `StructuredBuffer` is too limiting.

I don't know if we can use the same trick for constant buffers though IIRC they need to have a declared formats and a giant array of floats will have a stride of 16 bytes between each float.

**Comment by litherum** [comment]
Here's a proposal: add syntax, just like GLSL, to allow authors to optionally annotate a struct (block?) with `std140`, `std430`. Also, make the syntax for byte offsets on each field optional.

- If nothing is provided, default to something
- If the annotation is provided use it
- If the byte offsets are provided, use them
- If both the annotation and the byte offsets are provided, one of them wins

**Comment by magcius** [comment]
> The packing rules for DX10 / DX11 that you linked seem to be exactly the same as the `std140` rules.

No, they're not. In std140, `float` gets padded to `float2`, and `float3` gets padded to `float4` (assuming your vendor has implemented `std140` correctly. Some vendors pad everything to `float4`).

That means that:
```hlsl
cbuffer Light {
    float3 position;
    float alpha;
};
```

is 16 bytes in HLSL, but the closest GLSL equivalent (uniform blocks) would be 32 bytes in std140. This is why most shader authors pack everything into float4, often very awkwardly.

Note that arrays are not the same as vectors. `float position[3]` in HLSL is 48 bytes (arrays have their element types padded to 16 bytes, so `float position[3]` effectively becomes a float4 `position[3]` under the hood), but `float3 position` is 12 bytes.

Yes, padding and alignment is a mess, and a major compatibility hazard between shading languages and APIs. The ANGLE team is still dealing with vendor bugs about this.

**Comment by dj2** [comment]
Does MSL and HLSL do the same packing? If not, do we know how the WebGPU API will handle that? Is it going to shuffle data around in memory to fit the relevant platform?

**Comment by magcius** [comment]
> I've been wondering for a while if a smart enough compiler could translate scalar layout code into std140 / std430 by losing all typing information and looking at the buffer as a flat array of u32

Some hardware, especially vendors with vector ISAs, have fast load & swizzle capability. So you couldn't do `vec.yyzw` or similar with the same performance with this scheme.

**Comment by kainino0x** [comment]
Maybe this is annoying, but can we require `[[offset x]]` _and_ support an (optional or required, not sure) `[[layout std140]]` etc.? The spec would require that the layout and the explicit offsets are compatible.

My thinking is this makes it impossible for an implementation to mess up the layout silently - if they do, then valid code won't compile (and invalid code will).

To ease development, wgsl compilers could optionally say "layout and offsets didn't match. Did you mean to use the following offsets?" and developers can copy the suggested offsets into their source code.

**Comment by dj2** [comment]
That's an interesting idea. So, I'd do something like:

```
type b = [[layout std140]] struct {
  [[offset 0]] a : f32;
  [[offset 4]] b : f32;
};
```

Having the WGSL compiler tell me the offsets makes it a lot easier to figure out the offsets.

**Comment by jdashg** [comment]
It seems super verbose, but also it seems like the best idea for developers, for sure. It's gonna be much harder to mess up when you have to recite your knowledge of the (e.g. std140) binding rules explicitly!

I don't want to have to write all that, but also I'd have to know what these offsets are in when feeding in data anyway.

**Comment by grorg** [comment]
Discussed at [2020-03-17 meeting](https://docs.google.com/document/d/1WMQ_qRHM4R1BrrCm5Wdx_8ERdLem5jZG9kP65kXHC78/)

**Comment by litherum** [comment]
Another question that needs to be answered: If some structs require explicit offsets, which structs require them?

**Comment by grorg** [comment]
Resolution:

- group agreed to require explicit offsets on struct members
- this is already in the spec. @jdashg will check the content and send a PR if improvements are necessary
- might require a future issue on whether or not structures themselves require an attribute to describe the layout, which would give compiler a chance to tell the author they've made errors (as well as possibly help with some backend implementations)

**Comment by kvark** [comment]
Also, is it completely crazy to say: let's start with a single well-defined layout (that is the intersection of all the backend layouts expected) that we can spec, and then look into more flexible layouts as MVP-follow-ups or extensions? That way, we could have a simple web page, where the user can post their struct and get the offsets, in order to align their CPU structs.

Currently, the limitations on the explicit offsets depend on the usage of a struct. So we'd only really be able to report an error at the pipeline creation time. We can't tell if the offsets are wrong by just looking at the struct, or the shader.

**Comment by kainino0x** [comment]
It may already be the case that std430 is actually a superset of std140... I'm not sure.

Anyway, filed #621 about the layout decorations.

> Currently, the limitations on the explicit offsets depend on the usage of a struct. So we'd only really be able to report an error at the pipeline creation time. We can't tell if the offsets are wrong by just looking at the struct, or the shader.

Hmm, can't we? We can see what structs are used in the definition of a uniform input or a storage input/output. (Unless you're talking about a case with multiple shader modules for a single shader stage - in which case I'm not sure.)

**Comment by kvark** [comment]
> We can see what structs are used in the definition of a uniform input or a storage input/output.

Actually, I stand corrected. We can just observe all the uniform/storage structs and report if anything is wrong, regardless of the entry point usage. Thanks!

**Comment by dj2** [comment]
std430 and std140 differ around the layout of arrays and structures. 

For example, if you have something like `array<f32, 4>` in std140 that will take up 16 bytes per float so 64 bytes total. In std430 it will be 4 bytes per float so 16 bytes total.

**Comment by kainino0x** [comment]
Good point; I suppose it depends on exactly what we mean by superset. Would `[[layout std430, stride 16]] array<f32, 4>` be valid?

Even if not, I think the std140 layout:
```
type a = [[layout std140, stride 16]] array<f32, 4>;
```
could be represented in std430:
```
type myf32 = [[layout std430]] struct {
  [[offset 0]] x : f32;
  [[offset 12]] _pad : f32;
};
type b = [[layout std430, stride 16]] array<myf32, 4>;
```

**Comment by grorg** [comment]
Re-opening since there seems to be more discussion to happen here.

**Comment by kainino0x** [comment]
To clarify, none of this discussion was about including/excluding offset decorations. Just about what the layout rules would be. We should have the discussion elsewhere, possibly outside the [wgsl] subgroup because it affects both sides.

**Comment by litherum** [comment]
If I remember correctly, this issue was blocked on an investigation from @jdashg about what the specific rules would have to be.

**Comment by dneto0** [comment]
"What he specific rules have to be":  I take that as the policy side of things. The Offset is the mechanism side, which is already agreed to on March 17.

I suggest we have a different issue to track the policy.

I'll file another issue to get the complete picture for the mechanism, as it's missing ArrayStride at least.

Suggest we close this.

**Comment by johnkslang** [comment]
> There are, at least, four ways to specify the offsets in Vulkan. std140, std430,

Those are ways for OpenGL and GLSL, not Vulkan.  Vulkan uses explicit layout only, but then has different modes for what offsets are allowed.  But, it never strictly itself lays anything out.

>  add syntax, just like GLSL, to allow authors to optionally annotate a struct (block?) with std140 or std430

That is a known historical source of portability problems.

It would be much more portable for WGSL to just support explicit layout.  Maybe with the cross checking suggested by @kainino0x.

> It may already be the case that std430 is actually a superset of std140...

Yes, right, it allows more offsets, and does not take any away.

Two subtle points:

- What might be not intuitive is that thinking in terms of `std430 <-> offsets` is not bidirectional.  This is why GLSL "provides std430" while some Vulkan rules "are consistent with std430" (and similarly for std140).  For example, std430 can lay things out more tightly packed than std140, but a single set of offsets can satisfy both, and be wrong for neither.

- When a structure is used as a member type it two other structures, and those two other structures have different layout restrictions, where do you put the offsets for nested structure?  GLSL -> SPIR-V does this by starting with no offsets on the one shared nested structure, and then by the time it is in SPIR-V, there are two copies of the nested structure so that there are two places to decorate.

**Comment by dneto0** [comment]
Effectively we have resolved this in favour of what landed as https://github.com/gpuweb/gpuweb/pull/1447
